exchanges/orderbook/linked_list.go

```diff
@@ -0,0 +1,678 @@
+	for i := 0; i < len(items); i++ {
+		if *tip == nil {
+			// Extend node chain
+			*tip = stack.Pop()
+			// Set current node prev to last node
+			(*tip).prev = prev
+			ll.length++
+		}
+		// Set item value
+		(*tip).value = items[i]
+		// Set previous to current node
+		prev = *tip
+		// Set tip to next node
+		tip = &(*tip).next
+	}
+
+	// Push has references to dangling nodes that need to be removed and pushed
+	// back onto stack for re-use
+	var push *node
+	// Cleave unused reference chain from main chain
+	if prev == nil {
+		// The entire chain will need to be pushed back on to stack
+		push = *tip
+		ll.head = nil
+	} else {
+		push = prev.next
+		prev.next = nil
+	}
+
+	// Push unused pointers back on stack
+	for push != nil {
+		pending := push.next
+		stack.Push(push)
+		ll.length--
+		push = pending
+	}
+}
+
+// updateByID amends price by corresponding ID and returns an error if not found
+func (ll *linkedList) updateByID(updts []Item) error {
+updates:
+	for x := range updts {
+		for tip := ll.head; tip != nil; tip = tip.next {
+			if updts[x].ID != tip.value.ID { // Filter IDs that don't match
+				continue
+			}
+			tip.value.Price = updts[x].Price
+			tip.value.Amount = updts[x].Amount
+			continue updates
+		}
+		return fmt.Errorf("update error: %w %d not found",
+			errIDCannotBeMatched,
+			updts[x].ID)
+	}
+	return nil
+}
+
+// deleteByID deletes reference by ID
+func (ll *linkedList) deleteByID(updts Items, stack *stack, bypassErr bool) error {
+updates:
+	for x := range updts {
+		for tip := &ll.head; *tip != nil; tip = &(*tip).next {
+			if updts[x].ID != (*tip).value.ID {
+				continue
+			}
+			old := *tip
+			switch {
+			case old.prev == nil: // Tip is at head
+				// Shift everything to the left by setting the next node
+				*tip = old.next
+				// Dereference old node from current chain
+				if old.next != nil { // This is when liquidity hits zero
+					old.next.prev = nil
+				}
+			case old.next == nil: // Tip is at tail
+				// Remove old node
+				*tip = old.prev
+				old.prev.next = nil
+			default:
+				// Bypass old node with its prev and next
+				old.prev.next = old.next
+				old.next.prev = old.prev
+			}
+			stack.Push(old)
+			ll.length--
+			continue updates
+		}
+		if !bypassErr {
+			return fmt.Errorf("delete error: %w %d not found",
+				errIDCannotBeMatched,
+				updts[x].ID)
+		}
+	}
+	return nil
+}
+
+// cleanup reduces the max size of the depth length if exceeded. Is used after
+// updates have been applied instead of adhoc, reason being its easier to prune
+// at the end.
+func (ll *linkedList) cleanup(maxChainLength int, stack *stack) {
+	// Reduces the max length of total linked list chain, occurs after updates
+	// have been implemented as updates can push length out of bounds, if
+	// cleaved after that update, new update might not applied correctly.
+	if maxChainLength == 0 || ll.length <= maxChainLength {
+		return
+	}
+
+	n := ll.head
+	for i := 0; i < maxChainLength; i++ {
```

Comment:
I'm still trying to get my head around this stuff so I could be wrong but it seems like here we are going forward 1 too many times and then we do n.prev.next = nil. Instead we could just check if i == 0 { continue } after if n.next == nil check